### ISSUE TITLE
Tests fail when database is used at application bootstrap

### ISSUE THREAD

**Issue by warkadiuszz** [Codeception/module-yii2#98]
Hello,

after adding a function to the [bootstrap](https://www.yiiframework.com/doc/guide/2.0/en/structure-applications#bootstrap) of my Yii2 application, a lot of my tests stopped working. For all of them it was always one of two reasons:

1. `TooManyConnections` (just as the other old issues that happened in the past in that module, eg. [here](https://github.com/Codeception/Codeception/issues/1363) and [here](https://github.com/Codeception/Codeception/issues/4926)).
2. Data in the database not being cleared after each test (similar to [this](https://github.com/Codeception/Codeception/issues/4617))

All tests started working again immediately after removing the function in boostrap. Imporantly, the additional code in the boostrap uses database. For example (config `main.php`):
```
'bootstrap' => [
  'something' => function () {
    Model::find()->all();
  }
],
```

---

### Explanation / debugging / working fix

What I was able to find out is that both `ConnectionWatcher` and `TransactionForcer` classes work by [hooking to `Connection::EVENT_AFTER_OPEN` event](https://github.com/Codeception/module-yii2/blob/master/src/Codeception/Lib/Connector/Yii2/ConnectionWatcher.php#L38). When some code in `boostrap` uses database, before those two watchers are set up, the `EVENT_AFTER_OPEN` happens before them, so they never trigger.

The order of the code in `Module/Yii2::_before` ([here](https://github.com/Codeception/module-yii2/blob/master/src/Codeception/Module/Yii2.php#L331)) is:

```php
1: $this->client->startApp($this->yiiLogger);
 
2: $this->connectionWatcher = new Yii2Connector\ConnectionWatcher();
3: $this->connectionWatcher->start();

// ...

4: $this->startTransactions();
```

Application `boostrap` happens at step (1) - `startApp`. If database is used at that point, `Connection::EVENT_AFTER_OPEN` also happens. Then the handlers created by (2), (3) and (4) never work again(?), because they are waiting for that event to happen. 

If instead the `startApp` is called last:
``` php
2: $this->connectionWatcher = new Yii2Connector\ConnectionWatcher();
3: $this->connectionWatcher->start();

// ...

4: $this->startTransactions();
1: $this->client->startApp($this->yiiLogger);
```
Then both `ConnectionWatcher` and `TransactionForcer` (`startTransactions()`) bind to `Connection::EVENT_AFTER_OPEN` **before** bootstrap so they work correctly.

---

### Steps to reproduce

1. In application config add function that calls DB:
```php
'bootstrap' => [
  'something' => function () {
    SomeModel::find()->all();
  }
],
```

2. Create a test with 2 cases that relies on data being cleaned between them:
```php
<?php

namespace app\tests\unit;

use Codeception\Test\Unit;


class SomeTest extends Unit
{
    public function testCaseOne(): void
    {
	// Ensure no records are present in DB        
	$this->assertEquals(0, SomeModel::find()->count());
	
	// Create record
	$m = new SomeModel();
	$m->save();


	$this->assertEquals(1, SomeModel::find()->count());
    }

    public function testCaseTwo(): void
    {
	// Ensure no records are present in DB
	// This will fail if boostrap function using database is present
	$this->assertEquals(0, SomeModel::find()->count());
    }
}
```

---

With all that being said, I'm not sure if using database during application bootstrap is even legal. Offical Yii2 documentation does not seem to prohibit that though, at least I was not able to find such information. Also everything works.
Another thing is – I'm not sure if the order of those functions calls can be changed at all and if it won't be a breaking change for anyone else. It certainly works for me and, in my case, it has been broken since [this](https://github.com/Codeception/Codeception/pull/4894) rework from years ago. For now I keep that change in private fork but if the maintainers would be interested, I'll be happy to open a PR.

**Comment by samdark** [comment]
> I'm not sure if using database during application bootstrap is even legal

As you can see, that usage is tricky.

> Another thing is – I'm not sure if the order of those functions calls can be changed at all and if it won't be a breaking change for anyone else.

Likely it would break something. @SamMousa since you did that rework, do you remember order change reasoning?

**Comment by SamMousa** [comment]
I don't see an issue directly with changing the order. When do we remove class level event listeners? If we don't do that before starting the app in startApp it should be fine

**Comment by SamMousa** [comment]
@warkadiuszz I've created a fix here: https://github.com/Codeception/module-yii2/pull/133

Can you give this a go?

**Comment by warkadiuszz** [comment]
Hi, sorry for late response.

I tested it and unfortunately it doesn't seem to fix the problem.

After debugging it this time, I can tell now that from my initial report:
> 1. `TooManyConnections` (just as the other old issues that happened in the past in that module, eg. here and here).
> 2. Data in the database not being cleared after each test (similar to this)

In the newest version of the `module-yii2` the first one is actually fixable by setting `PDO::ATTR_PERSISTENT => true`. So it's not a problem anymore. (It might have been like that even back then).

Only the second issue is still there – that the database is not cleared in between the tests. And unfortunately the patch from #133 does not resolve that. I tested it in my application and in clean empty project from `composer create-project` and still reproduced the original issue. 

I also tested if the hotfix that I came up with last year and used so far works with the newest version of `module-yii2` and even with your patch applied. And it does – so still moving `$this->getClient->startApp(..)` to the end of `Module/Yii2::_before()` fixes the issue. But I cannot explain how and why, I don't know the internal mechanisms well enough.

---

I have that minimal app created from the template where the issue is reproducible, would you like me to quickly put it in some repo for you to easily test it yourself?

**Comment by warkadiuszz** [comment]
Hello

I was able to spend some more time debugging the issue and have some information that might be helpful. Sorry for the amount of text.

tl;dr: Opening database at bootstrap causes `ConnectionWatcher` and `TransactionForcer` to miss the connection and not work correctly. Possible fix at the end.

## The problem

So turns out that using DB at bootstrap causes really two separate issues:
1. "Leaking" DB connections that are not caught by `ConnectionWatcher` and cause `Too many open connections` error
2. `TransactionForcer` missing the database connection because the connection is opened before `TransactionForcer` is started. Prevents data from being cleared (transactions rolled back) between tests. 

## Execution flow without DB access in bootstrap

When database connection is not opened in `bootstrap`, the normal execution is as follows:
1. Application is started ([`Module/Yii2.php:394`](https://github.com/Codeception/module-yii2/blob/master/src/Codeception/Module/Yii2.php#L394)
2. `ConnectionWatcher` is created and `->start()`ed - it binds to `Connection::EVENT_AFTER_OPEN` and waits for the DB connection to open ([`Module/Yii2.php:396`](https://github.com/Codeception/module-yii2/blob/master/src/Codeception/Module/Yii2.php#L396))
3. (optionally fixtures are loaded, I have not tested that)
4. `TransactionForcer` is created, also binds to `Connection::EVENT_AFTER_OPEN` and waits for DB connection to open too ([`Module/Yii2:474`](https://github.com/Codeception/module-yii2/blob/master/src/Codeception/Module/Yii2.php#L474))

Then the tests start execution, application code performs first operation with the database so the `Connection` is open. That triggers `Connection::EVENT_AFTER_OPEN`, `TransactionForcer::connectionOpened()` is run and starts transaction. After the test `TransactionForcer::rollbackAll()` is called and the cycle repeats with the next test. Everything works correctly.

## Execution flow with DB access at bootstrap (buggy)

With that in mind, it's easier to understand why opening the database connection before the test body is executed causes issues. In that case, the execution is:

1. Application is started ([`Module/Yii2.php:394`](https://github.com/Codeception/module-yii2/blob/master/src/Codeception/Module/Yii2.php#L394)
2. During application start, the database connection is open - since database is accessed in `bootstrap`. This happens in [`Connector/Yii2.php:309`](https://github.com/Codeception/module-yii2/blob/master/src/Codeception/Lib/Connector/Yii2.php#L309). In normal execution there is no database connection opened at this time. This triggers `Connection::EVENT_AFTER_OPEN` already, but nothing binds to it yet.
3. `ConnectionWatcher` is created and `->start()`ed - it binds to `Connection::EVENT_AFTER_OPEN` and waits for the DB connection to open ([`Module/Yii2.php:396`](https://github.com/Codeception/module-yii2/blob/master/src/Codeception/Module/Yii2.php#L396)) - but it has been already opened.
4. (optionally fixtures are loaded, I have not tested that)
5. `TransactionForcer` is created, also binds to `Connection::EVENT_AFTER_OPEN` and waits for DB connection to open too ([`Module/Yii2:474`](https://github.com/Codeception/module-yii2/blob/master/src/Codeception/Module/Yii2.php#L474)). As with `ConnectionWatcher` in step 3., the event has been triggered before that, so `TransactionForcer` never catches DB connection, never rolls back any transactions.

---

## Working fix

Now, when it comes to the fix, this is where I got.

1. `ConnectionWatcher` has to be started right before `->startApp()`. 

Your fix in https://github.com/Codeception/module-yii2/pull/133 partially addresses that. The issue there that remains is that `Client/Yii2::resetApplication` (called [here](https://github.com/Codeception/module-yii2/blob/master/src/Codeception/Module/Yii2.php#L377)) calls `Event::offAll()` which unbinds the `ConnectionWatcher` from `Connection::EVENT_AFTER_OPEN` and it doesn't see any further connections being open.

How I fix that in my patch is to call additional `$this->connnectionWatcher->start();` right before `->startApp()`. See [here](https://github.com/baningo/module-yii2/blob/fix/98-tests-fail-when-db-is-used-at-bootstrap/src/Codeception/Module/Yii2.php#L399). This ensures that the `ConnectionWatcher` will always catch the `Connection::EVENT_AFTER_OPEN`.

This solves the issue with `TooManyConnections` because `ConnectionWatcher` manages them correctly, nothing slips through. But because database is already open, Yii reuses the connection and another `Connection:EVENT_AFTER_OPEN` is not called when the test body starts (and application/test code accesses DB) so `TransactionForcer` does not see the connection and does not start any transactions. To fix that see 2. below.

2. Fresh DB connection has to be open at the beginning of each test body.

Because the database connection is already open when the test body gets executed, `TransactionForcer` does not wrap the test body with transaction. Simply closing the database connection (`Yii::$app->db->close()`) at the end of my `bootstrap` code solved the issue immediately. This creates the following execution flow:
1. `ConnectionWatcher` is created and started
2. `->startApp()` is called, `bootstrap` opens DB connection, closes it quickly
3. `TransactionForcer` is created and started, binds to `Connection::EVENT_AFTER_OPEN`
4. Test body gets executed, application/test code tries to access the DB
5. Since DB was closed before, a new connection is open, `EVENT_AFTER_OPEN` is triggered, `TransactionForcer` catches it and starts transaction.

To avoid modifying my bootstrap code (and make the fix potentially upstreamable) I call `Yii::$app->db->close()` after creating the app in `Client\Yii2::startApp`, see [here](https://github.com/baningo/module-yii2/blob/fix/98-tests-fail-when-db-is-used-at-bootstrap/src/Codeception/Lib/Connector/Yii2.php#L314). Probably needs some checks (is db component there, is connection open etc.) but in this simplest form, it solved the issue for me.

---

So, at the end it seems like two interwined bugs interacting at the same time. I'm not sure if my patch can directly be applied here but maybe at least it can inspire proper fix. Let me know if you'd like me to polish what's in https://github.com/baningo/module-yii2/tree/fix/98-tests-fail-when-db-is-used-at-bootstrap and open PR here or if I can help in any other way.

**Comment by warkadiuszz** [comment]
> After reading your description I'm even more convinced that not accessing database in bootstrap is the way to go

This is absolutely fine by me. Since not so many people seem to have that problem, I'll just keep using my patch. Feel free to close the issue any time.

>  I'm not excited to add more workaround code to fix edge cases.

Understandable

> Could you elaborate on why database access during bootstrapping is needed? I'm curious!

In a nutshell, in production code we fetch "customer" (tenant) record from DB based on the domain used in the request (eg. `customer1.example.org`). We use it to configure our custom, global `Customer` component that is then used throughout the application code (incl. other bootstraped components...) to get details about the customer.
All our code expects that component to be there and be configured with the information from the database. So do tests. There is couple more complications that prevent the first DB access from being moved to eg. aforementioned `Connection::EVENT_AFTER_OPEN` or higher up the stream. 

Anyway, thank you for your time developing the module, investigating the issue and even coming up with potential fix.

**Comment by SamMousa** [comment]
We have the same scenario in our application!

I do this processing in before action, because it is dependent on the request object (which you should also not access in application Bootstrap).

Imagine one test body that does requests to 2 hostnames, that won't work if set up via bootstrap.

I find it a challenge to have component inter dependency in a clean way in Yii2, because you can't do constructor injection for components cleanly...

I'll think about this some more...

**Comment by SamMousa** [comment]
Okay, I dove into our code for a bit.
The solution we settled on:
1. Have a base controller for with a property (in our case `public Organisation $organisation;`)
2. Have a behavior that loads this property; approximately like this:

```php
/**
 * @property Controller $owner
 */
final class LoadPanelFromHostBehavior extends Behavior
{
    public function events(): array
    {
        return [
            Controller::EVENT_BEFORE_ACTION => $this->beforeAction(...)
        ];
    }

    public function attach($owner): void
    {
        parent::attach($owner);
        if (!$owner instanceof Controller) {
            throw new InvalidArgumentException('This behavior must be attached to a PanelMember controller');
        }
    }


    private function beforeAction(): void
    {
        $request = $this->owner->request;
        if (!$request instanceof Request) {
            throw new InvalidConfigException('Request must be a web request');
        }
        $hostName = $request->getHostName();
        if (!is_string($hostName)) {
            throw new BadRequestHttpException('Hostname must be set');
        }
        $this->owner->panel = $this->findPanelByHost($hostName);
    }

    private function findPanelByHost(string $host): Panel
    {
        $result = Panel::find()
            ->andWhere(['host' => $host])
            ->model();
        if (!isset($result)) {
            throw new BadRequestHttpException("Panel {$host} not found");
        }

        \Yii::$app->language = $result->locale;

        return $result;
    }


```

Not sure if useful, this could probably also be adapted to update / configure your service instead.

**Comment by warkadiuszz** [comment]
Thank you very much for the code sample. I'll attempt to adapt it to our system, it makes more sense than dealing with the bootstrap gray-area.
